fe/src/main/java/org/apache/doris/planner/MaterializedViewSelector.java

```diff
@@ -0,0 +1,453 @@
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * The new materialized view selector supports SPJ<->SPJG.
+ * At the same time, it is compatible with all the features of the old version.
+ * The SPJ query is "Select Projection and Join" such as:
+ *     select t1.c1 from t1, t2 where t1.c2=t2.c2 and t1.c3=1;
+ * The SPJG query is "Select Projection Join and Group-by" such as:
+ *     select t1.c1, sum(t2.c1) from t1, t2 where t1.c2=t2.c2 and t1.c3=1 group by t1.c1;
+ */
+public class MaterializedViewSelector {
+    private static final Logger LOG = LogManager.getLogger(MaterializedViewSelector.class);
+
+    private final SelectStmt selectStmt;
+    private final Analyzer analyzer;
+
+    private Map<String, Set<String>> columnNamesInPredicates = Maps.newHashMap();
+    private boolean isSPJQuery;
+    private Map<String, Set<String>> columnNamesInGrouping = Maps.newHashMap();
+    private Map<String, Set<AggregatedColumn>> aggregateColumnsInQuery = Maps.newHashMap();
+    private Map<String, Set<String>> columnNamesInQueryOutput = Maps.newHashMap();
+
+    private boolean disableSPJGView;
+    private String reasonOfDisable;
+    private boolean isPreAggregation = true;
+
+    public MaterializedViewSelector(SelectStmt selectStmt, Analyzer analyzer) {
+        this.selectStmt = selectStmt;
+        this.analyzer = analyzer;
+        init();
+    }
+
+    /**
+     * There are two stages to choosing the best MV.
+     * Phase 1: Predicates
+     * According to aggregation and column information in the select stmt,
+     * the candidate MVs that meets the query conditions are selected.
+     * Phase 2: Priorities
+     * According to prefix index and row count in candidate MVs,
+     * the best MV is selected.
+     *
+     * @param scanNode
+     * @return
+     */
+    public void selectBestMV(ScanNode scanNode) throws UserException {
+        long start = System.currentTimeMillis();
+        Preconditions.checkState(scanNode instanceof OlapScanNode);
+        OlapScanNode olapScanNode = (OlapScanNode) scanNode;
+        Map<Long, List<Column>> candidateIndexIdToSchema = predicates(olapScanNode);
+        long bestIndexId = priorities(olapScanNode, candidateIndexIdToSchema);
+        LOG.info("The best materialized view is {} for scan node {} in query {}, cost {}",
+                 bestIndexId, scanNode.getId(), selectStmt.toSql(), (System.currentTimeMillis() - start));
+        olapScanNode.updateScanRangeInfo(bestIndexId, isPreAggregation, reasonOfDisable);
+    }
+
+    private Map<Long, List<Column>> predicates(OlapScanNode scanNode) {
+        // Step1: all of predicates is compensating predicates
```

Comment:
What is the meaning of `compensating predicates` here?  Is it same with `compensating predicates` in `Optimizing Queries Using Materialized Views:A Practical, Scalable Solution` paper?

Reply:
Yes, it same as the compensating predicates in Optimizing Queries Using Materialized Views:A Practical, Scalable Solution paper.
In the present, doris could not support the materialized view which has predicates.
So all of predicates in query is compensating predicates directly.

Reply:
OK. I see.